### ISSUE TITLE
[spike] Data tables

### ISSUE THREAD

**Issue by danswick** [GSA-TTS/FAC#3874]
# Trying out Data Tables:

[DataTables](https://datatables.net/ ) is a Javascript library that allows for advanced table interaction with not-that-much effort. It includes features like filtering, searching, column selection, and CSV/XLSX downloads. 

But as a JS library, it doesn't work nicely out of the box with Django.

[django-ajax-datatable](https://pypi.org/project/django-ajax-datatable/) allows one to serve a DataTable with all of its features against a Django model. It allows for server-side processing, such that one can select columns and apply filters in real time without a page reload. 

# Outcomes?

1. Can DataTables work with Django?
    - Can it be installed/enabled without breaking things?
    - Will it involve a reasonable amount of effort to maintain? 
    - Are required & optional libraries kept up-to-date?
2. Is it performant on the amount of data we work with regularly (>300k general objects)?
    - Does the reactive element of django-ajax-datatables lock up the db/app? Or, are requests so frequent when searching/filtering that it locks things up for other users?
    - Remembering that the local stack will almost always be faster than the online environments
    - Maybe have a test on Preview
4. Will DataTables solve problems our users have?
    - Does it go beyond "hey it's a different way to do excel downloads?"
    - Is column selection usable & intuitive on data with this many total fields?
    - Can we make use of the materialized view (DisseminationCombined) to serve award/finding information alongside the general objects?

**Comment by danswick** [comment]
A suggestion for the "is it performant" category: how will multiple DataTables users affect others' ability to use the app? Will intensive db reads lock up **the db** for other users? Can an **app instance** get swamped processing multiple large requests?

**Comment by jadudm** [comment]
A note about the performance piece (which does not mitigate any of the other concerns):

When I demoed this, I pulled 30K records via API, and then all operations were in memory. A solution that fires multiple queries to the DB (e.g. on every keypress in a search field) would not be our preferred approach; sending off a query, and then returning *n*K results, and allowing DataTables to work with it in memory might be.

The spike answered good questions. Regardless of we come back to this tech or not:

* Faster pagination/sorting is desirable
* Export of all tables as CSV or XLSX is desirable
* Customization/selection of columns (from the big `VIEW` table) is desirable

These are not yet storied, but perhaps we can story the exploration from a user perspective in the next sprint. That might guide further exploration.

We can leave this open for further convo, or close, whichever makes sense to you, @danswick .
